pkg/plugins/config-gen/v1alpha/scaffolds/internal/templates/makefile.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+)
+
+var _ machinery.Template = &Makefile{}
+
+// Makefile scaffolds a file that defines project management CLI commands
+type Makefile struct {
+	machinery.TemplateMixin
+	machinery.ComponentConfigMixin
+
+	// Image is controller manager image name
+	Image string
+	// BoilerplatePath is the path to the boilerplate file
+	BoilerplatePath string
+	// Controller tools version to use in the project
+	ControllerToolsVersion string
+	// Kustomize version to use in the project
+	KustomizeVersion string
+	// ControllerRuntimeVersion version to be used to download the envtest setup script
+	ControllerRuntimeVersion string
+}
+
+// SetTemplateDefaults implements file.Template
+func (f *Makefile) SetTemplateDefaults() error {
+	if f.Path == "" {
+		f.Path = "Makefile"
+	}
+
+	f.TemplateBody = makefileTemplate
+
+	if f.Image == "" {
+		f.Image = "controller:latest"
+	}
+
+	// FIXME: How can we customize makefile without overwrite it since it is language based?
+	f.IfExistsAction = machinery.OverwriteFile
```

Comment:
Need some help here:
This `config-gen` plugin will generate a makefile that might overwrite the one from `go/v3`.  So there are two concerns:
1. Most part of the template from both are duplicate. (The main difference is just `make deploy` and `make install`). 
2. It potentially leads contradiction if another plugin is used to generate makefile.

Reply:
I suppose we may expect this plugin to be a more general one.(So to make it available in sdk)

I will get some references on the plugins from sdk to see how we can adjust this.

Reply:
The Makefile is part of the language plugin.
See that we have the kustomize and the base.go/v3 which are a composition to generate the = go/v3 

IMO:
How it will works? We will need to call base.go.<version > + config-gen such as we do now for go/v3: https://github.com/kubernetes-sigs/kubebuilder/blob/master/cmd/main.go#L35-L39

So, why does config-gen needs to scaffold the Makefile? The makefile is scaffolded by the Golang one because it has specific targets which are related to the language ( see that the kustomize plugin is used with other languages such as in SDK with Helm/Ansible so that their specific language plugins do not have go commands on the targets ) 


More info: See the plugin doc PR : https://github.com/kubernetes-sigs/kubebuilder/pull/2569

Reply:
Please see: https://github.com/kubernetes-sigs/kubebuilder/pull/2540/files#r841070549